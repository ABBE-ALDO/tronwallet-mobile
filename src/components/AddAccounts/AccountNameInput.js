import React, { PureComponent, Fragment } from 'react'
import { string, number, func } from 'prop-types'

import { isNameValid, isAliasUnique } from '../../utils/validations'
import { formatAlias } from '../../utils/contactUtils'
import tl from '../../utils/i18n'

import Input from '../Input'
import { Colors } from '../DesignSystem'
import { ErrorText } from '../AddressBook/AddressForm/elements'

class AccountNameInput extends PureComponent {
  static propTypes = {
    pin: string.isRequired,
    totalAccounts: number.isRequired,
    onChangeText: func
  }

  static defaultProps = {
    onChangeText: () => {}
  }

  state = {
    accountName: '',
    accountNameError: null
  }

  static getDerivedStateFromProps (props, state) {
    const { totalAccounts } = props

    if (state.prevTotalAccounts !== totalAccounts) {
      const accountName = `Account ${totalAccounts}`

      return {
        accountName,
        prevTotalAccounts: totalAccounts
      }
    }

    return null
  }

  _validateAccountName = async name => {
    if (name) {
      if (!isNameValid(name)) {
        return tl.t('addressBook.form.nameError')
      }

      const aliasIsUnique = await isAliasUnique(formatAlias(name), this.props.pin)
      if (!aliasIsUnique) {
        return tl.t('addressBook.form.uniqueAliasError')
      }
    }

    return null
  }

  _handleAccountNameChange = async accountName => {
    const accountNameError = await this._validateAccountName(accountName)

    this.setState({ accountName, accountNameError })
    this.props.onChangeText(accountName, accountNameError)
  }

  render () {
    const { accountName, accountNameError } = this.state

    return (
      <Fragment>
        <Input
          label={tl.t('addressBook.form.name')}
          value={accountName}
          onChangeText={this._handleAccountNameChange}
          placeholder={tl.t('newAccount.placeholder')}
          labelWrapperColor={Colors.dusk}
        />
        {!!accountNameError && (<ErrorText>{accountNameError}</ErrorText>)}
      </Fragment>
    )
  }
}

export default AccountNameInput
