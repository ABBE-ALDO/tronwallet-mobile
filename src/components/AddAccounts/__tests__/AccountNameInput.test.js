import React from 'react'
import { shallow } from 'enzyme'

import AccountNameInput from '../AccountNameInput'

jest.mock('../../../utils/i18n')

describe('Account name input component', () => {
  const setup = propOverrides => {
    const props = Object.assign({
      pin: '123456',
      totalAccounts: 1
    }, propOverrides)

    const wrapper = shallow(
      <AccountNameInput {...props} />
    )

    return {
      props,
      wrapper
    }
  }

  test('Should match when render component', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
  })

  test('Should suggest account name when set total account', () => {
    const { wrapper } = setup()
    const { accountName } = wrapper.state()
    expect(accountName).toBe('Account 1')
  })
})
