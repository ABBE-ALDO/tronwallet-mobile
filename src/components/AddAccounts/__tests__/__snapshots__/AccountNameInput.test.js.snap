// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account name input component Should match when render component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AccountNameInput
    onChangeText={[Function]}
    pin="123456"
    totalAccounts={1}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Input
          editable={true}
          label="NAME"
          labelWrapperColor="#3f415d"
          numbersOnly={false}
          onChangeText={[Function]}
          placeholder="My precious account"
          placeholderTextColor="#66688F"
          returnKeyType="send"
          showPlaceholder={true}
          value="Account 1"
        />,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "editable": true,
          "label": "NAME",
          "labelWrapperColor": "#3f415d",
          "numbersOnly": false,
          "onChangeText": [Function],
          "placeholder": "My precious account",
          "placeholderTextColor": "#66688F",
          "returnKeyType": "send",
          "showPlaceholder": true,
          "value": "Account 1",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Input
            editable={true}
            label="NAME"
            labelWrapperColor="#3f415d"
            numbersOnly={false}
            onChangeText={[Function]}
            placeholder="My precious account"
            placeholderTextColor="#66688F"
            returnKeyType="send"
            showPlaceholder={true}
            value="Account 1"
          />,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "editable": true,
            "label": "NAME",
            "labelWrapperColor": "#3f415d",
            "numbersOnly": false,
            "onChangeText": [Function],
            "placeholder": "My precious account",
            "placeholderTextColor": "#66688F",
            "returnKeyType": "send",
            "showPlaceholder": true,
            "value": "Account 1",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenThreeAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
